### PR TITLE
[Scheduler] Add re-routing to chief 

### PR DESCRIPTION
**Pull request by Tankilevitch**
This PR, is part of a series of PRs making the API stateless. 
We decided that only the chief will be responsible for running the scheduler, which means that any code that had any effect on the scheduler behavior (e.g create / delete / update / invoke ) must run by the chief. 
To implement that requirement, when accessing code that affect the scheduler behavior we will check which instance type we are running inside, and if we are running inside a worker we will re-route the request to the chief.

### Schedules:
- CREATE schedule- `/projects/{project}/schedules`
- PUT schedule - `/projects/{project}/schedules/{name}`
- DELETE schedule - `/projects/{project}/schedules/{name}`
- DELETE project schedules - `/projects/{project}/schedules`
- POST schedule - - `/projects/{project}/schedules/{name}/invoke`

We also had endpoints which weren't explicitly related to scheduling but had some features that if enabled they would use the scheduler.
### Submit Job
- CREATE  job - `submit_job` : when contains `schedule` attribute then re-route to chief.

### Functions
- CREATE build function - `/build/function` : when contains `track_models` attribute then re-route to chief.

### Projects
- DELETE project - `/projects/{name}`: as part of deleting project we are also deleting related resources, one of them is schedules, to make sure we are removing them from the chief, we are re-routing to chief.

I've also added a decorator `ensure_function_running_on_chief_mode` for functions that supposed to run only on chief, hopefully this could help us fail early when developing and using existing functions that supposed to run only on chief.

### CODE REVIEW

File: mlrun/api/utils/background_tasks.py [hunk 1]
```diff
@@ -148,6 +152,7 @@ async def background_task_wrapper(self, name: str, function, *args, **kwargs):
                 name, mlrun.api.schemas.BackgroundTaskState.succeeded
             )
 
+    @mlrun.api.utils.helpers.run_on_chief_only
     def _update_background_task(
```
Comment (Hedingber) [inline mlrun/api/utils/background_tasks.py:156]: I don't think it makes sense to add the decorator on private functions, public ones should be enough (relevant in other places as well)

File: mlrun/api/utils/clients/chief.py [hunk 2]
```diff
@@ -48,23 +48,87 @@ def get_background_task(
...
     def trigger_migrations(self, request: fastapi.Request = None) -> fastapi.Response:
         request_kwargs = self._resolve_request_kwargs_from_request(request)
         return self._proxy_request_to_chief(
             method="POST", path="operations/migrations", **request_kwargs
         )
 
+    def create_schedule(
+        self, project: str, request: fastapi.Request
+    ) -> fastapi.Response:
+        request_kwargs = self._resolve_request_kwargs_from_request(request)
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/api/utils/clients/chief.py:60]: I understand it means we'll have `_proxy_request_to_chief` and `_proxy_request_to_chief_async` - sounds reasonable

File: mlrun/api/utils/helpers.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        return function(*args, **kwargs)
+
+    async def async_wrapper(*args, **kwargs):
+        return await function(*args, **kwargs)
+
+    if (
+        mlrun.mlconf.httpdb.clusterization.role
+        != mlrun.api.schemas.ClusterizationRole.chief
+    ):
+        if mlrun.mlconf.fail_on_running_chief_functions_in_worker_mode == "enabled":
```
Comment (Hedingber) [inline mlrun/api/utils/helpers.py:27]: Great thinking for making this configurable 🥇

File: tests/api/api/test_submit.py [hunk 4]
```diff
@@ -215,6 +213,120 @@ def test_submit_job_service_accounts(
...
+    monkeypatch.setattr(
+        mlrun.api.utils.clients.chief,
+        "Client",
+        lambda *args, **kwargs: handler_mock,
+    )
+
+    submit_job_body = _create_submit_job_body_with_schedule(function, project)
+    json_body = mlrun.utils.dict_to_json(submit_job_body)
+    client.post("submit_job", data=json_body)
+    assert handler_mock._proxy_request_to_chief.call_count == 1
```
Comment (Hedingber) [inline tests/api/api/test_submit.py:250]: same

**Review by Hedingber — CHANGES_REQUESTED** [review]
Looks good overall, pretty minor comments
Amazing test coverage, bravo for doing the import changes

File: mlrun/api/utils/background_tasks.py [hunk 5]
```diff
@@ -95,6 +96,7 @@ def __init__(self):
             str, mlrun.api.schemas.BackgroundTask
         ] = {}
 
+    @mlrun.api.utils.helpers.run_on_chief_only
```
[the remark this answers is not in the document]
Comment (Tankilevitch) [reply mlrun/api/utils/background_tasks.py:99]: right, ensure sounds better, changed.

File: tests/api/api/test_functions.py [hunk 6]
```diff
@@ -109,7 +114,86 @@ async def test_multiple_store_function_race_condition(
...
+    json_body = _generate_build_function_request(function)
+    client.post(endpoint, data=json_body)
+    # no schedule inside job body, expecting to be run in worker
+    assert handler_mock._proxy_request_to_chief.call_count == 0
+
+    function_with_track_models = _generate_function(function_name)
+    function_with_track_models.spec.track_models = True
+    json_body = _generate_build_function_request(function_with_track_models)
+    client.post(endpoint, data=json_body)
+    assert handler_mock._proxy_request_to_chief.call_count == 1
```
[the remark this answers is not in the document]
Comment (Tankilevitch) [reply tests/api/api/test_functions.py:149]: I am not testing whether the chief can handle the request in this test, the one bellow tests that.
This test purpose it to test the flow logic is correct.

File: tests/api/api/test_submit.py (hunk 4 above)
Comment (Tankilevitch) [reply tests/api/api/test_submit.py:250]: same ^.

File: mlrun/api/utils/clients/chief.py [hunk 7]
```diff
@@ -78,9 +142,16 @@ def _resolve_request_kwargs_from_request(
...
+        """
+        try:
+            loop = asyncio.get_event_loop()
+        except RuntimeError as exc:
+            if "There is no current event loop in thread" in str(exc):
+                loop = asyncio.new_event_loop()
+                asyncio.set_event_loop(loop)
         future = asyncio.ensure_future(request.body())
         loop.run_until_complete(future)
         return future.result()
```
[the remark this answers is not in the document]
Comment (Tankilevitch) [reply mlrun/api/utils/clients/chief.py:157]: Changed so the caller will pass the body, no need for using event loops any more.

**Pull request merged**